src/fal/dag.py

```diff
@@ -0,0 +1,129 @@
+from pathlib import Path
+import re
+
+T = TypeVar("T", bound="FalScript")
+
+@dataclass(frozen=True)
+class FalScript:
+    model: DbtModel
+    path: str
+    dependencies: List[T] = field(default_factory=list)
+
+    def __hash__(self):
+        return self.path.__hash__() * self.model.name.__hash__()
+
+    def __repr__(self):
+        return "FalScript(" + self.path.__str__() + ")"
+
+    def parse(self) -> List[str]:
+        """
+        Parses the python file to identify any dependencies
+        """
+        with open(self.path) as file:
+            lines = file.readlines()
+            ## TODO: do this search in parallel
+            dependency_names = []
+            for line in lines:
+                match = re.search('ref\("(.*)"\)', line)
+                if match:
+                    group = match.group(1)
+                    dependency_names.append(group)
+            return dependency_names    
+    
+    def addDependency(self, dependency: T):
+        if (dependency not in self.dependencies):
+            self.dependencies.append(dependency)
+
+    def exec(self, ref, context, source):
+        """
+        Executes the script
+        """
+        with open(path) as file:
+            a_script = file.read()
+            exec(
+                a_script,
+                {
+                    "ref": ref,
+                    "context": context,
+                    "source": source,
+                },
+            )
+
+@dataclass(frozen=True)
+class UniqueKey:
+    model: DbtModel
+    script_path: Path
+
+class ScriptGraphBuilder:
+    modelNameToModelLookup: Dict[str, DbtModel] = {}
+    modelToScriptLookup: Dict[str, List[Path]] = {}
+    falScripts: Dict[UniqueKey, FalScript] = {}
+
+    def __init__(self, models: List[DbtModel], keyword: str, root: str):
+        for model in models:
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        for model_name, script_list in self.modelToScriptLookup.items():
+            for script in script_list:
+                self.recursively_set_dependencies(UniqueKey(self.modelNameToModelLookup[model_name], script))
+
+    def recursively_set_dependencies(self, key: UniqueKey):
+        if key in self.falScripts:
+          return
+        self.falScripts[key] = FalScript(key.model, key.script_path)
+        current_script = self.falScripts.get(key)
+        dependency_models = current_script.parse()
+       
+        for model_name in dependency_models:
+            model = self.modelNameToModelLookup[model_name]
+            dependency_scripts = self.modelToScriptLookup.get(model_name, [])
+
+            for dependency_script in dependency_scripts:
+                dependency_key = UniqueKey(model, dependency_script)
+                if (self.falScripts.get(dependency_key) is None):
+                    self.recursively_set_dependencies(dependency_key)
```

Comment:
You can remove the `if` and call directly. It will be handled by the line 75 when called